.NET 10 SDK ARM64: Illegal instruction (SIGILL) on Apple M4 with macOS Virtualization.Framework

### Description

When building .NET 10 applications in Docker on Apple M4 hardware, the .NET 10 SDK (version 10.0.101) crashes with "Illegal instruction (core dumped)" on ARM64 images. The crash occurs intermittently during various .NET CLI operations including `dotnet new`, `dotnet add package`, and `dotnet build`.

**Environment:**
- **Host**: macOS (Sequoia) on Apple M4 Pro
- **Virtualization**: macOS Virtualization.Framework (VZ) via Colima
  - Colima config: `vmType: vz`, `arch: aarch64`, `rosetta: false`
- **VM OS**: Ubuntu 24.04.1 LTS (Noble Numbat)
- **VM Architecture**: aarch64 (native ARM64, not emulated)
- **.NET SDK**: 10.0.101
- **Docker Image**: `mcr.microsoft.com/dotnet/sdk:10.0` (ARM64 variant, sha256:d1823fecac...)

**Key Observations:**
- ✅ Works perfectly when using `--platform linux/amd64` (via Rosetta translation)
- ❌ Fails on native ARM64 (`--platform linux/arm64`)
- The bug is intermittent but consistently reproducible with `--no-cache` builds
- Multi-project solutions with references trigger it more reliably than single projects
- Exit code: 132 (SIGILL - Illegal Instruction)

**Impact:**
Cannot build .NET 10 applications natively on ARM64 in macOS VZ environment. Workaround is to build for AMD64 and rely on Rosetta, but this defeats the purpose of native ARM64 support.

### Reproduction Steps

```
docker run --rm --platform linux/arm64 mcr.microsoft.com/dotnet/sdk:10.0 \
  sh -c 'dotnet new classlib -n Test'
```

The above fails intermittently but regularly, see sample output below:

```
> docker run --rm --platform linux/arm64 mcr.microsoft.com/dotnet/sdk:10.0 \
  sh -c 'dotnet new classlib -n Test'
The template "Class Library" was created successfully.

Processing post-creation actions...
Restoring /Test/Test.csproj:
  Determining projects to restore...
  Restored /Test/Test.csproj (in 30 ms).
Restore succeeded.


> docker run --rm --platform linux/arm64 mcr.microsoft.com/dotnet/sdk:10.0 \
  sh -c 'dotnet new classlib -n Test'
Illegal instruction (core dumped)
> docker run --rm --platform linux/arm64 mcr.microsoft.com/dotnet/sdk:10.0 \
  sh -c 'dotnet new classlib -n Test'
The template "Class Library" was created successfully.

Processing post-creation actions...
Restoring /Test/Test.csproj:
  Determining projects to restore...
  Restored /Test/Test.csproj (in 33 ms).
Restore succeeded.


> docker run --rm --platform linux/arm64 mcr.microsoft.com/dotnet/sdk:10.0 \
  sh -c 'dotnet new classlib -n Test'
Illegal instruction (core dumped)
```

### Expected behavior

Build succeeds

### Actual behavior

Build fails (intermittently)

### Regression?

_No response_

### Known Workarounds

Build an AMD64 image instead

### Configuration

Using SDK 10.0

### Other information

### System Information

```bash
# Host CPU
$ sysctl machdep.cpu.brand_string
machdep.cpu.brand_string: Apple M4 Pro

# Docker/VM architecture
$ docker info | grep Architecture
Architecture: aarch64

# .NET SDK version in container
$ docker run --rm mcr.microsoft.com/dotnet/sdk:10.0 dotnet --version
10.0.101

# CPU features visible in container
$ docker run --rm mcr.microsoft.com/dotnet/sdk:10.0 cat /proc/cpuinfo | grep Features
Features: fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 asimddp sha512 asimdfhm dit uscat ilrcpc flagm sb paca pacg dcpodp sve2 flagm2 frint svei8mm svebf16 bf16 afp sme smei16i64 smef64f64 smei8i32 smef16f32 smeb16f32 smef32f32 sme2 smei16i32 smebi32i32
```

I had a very similar error when building an image with Podman, running `dotnet build` from inside the container on a Mac with M4.

I always saw one of two different errors. Which one i got seemed very random:
- error MSB6006: "csc.dll" exited with code 132. 
- Illegal instruction

I dont know what the underlying problem is, but i found a workaround by making sure i create the Podman Machine with LibKrun instead of Apple HyperVisor:

<img width="1519" height="786" alt="Image" src="https://github.com/user-attachments/assets/8b949029-e569-4387-98fe-162d69910132" />

--- a/src/coreclr/pal/src/arch/arm64/context2.S
+++ b/src/coreclr/pal/src/arch/arm64/context2.S
@@ -296,11 +296,3 @@ LEAF_END RestoreCompleteContext, _TEXT
 
 #endif // __APPLE__
 
-// Incoming:
-//  None
-//
-.arch_extension sve
-    LEAF_ENTRY CONTEXT_GetSveLengthFromOS, _TEXT
-        rdvl    x0, 1
-        ret     lr
-    LEAF_END CONTEXT_GetSveLengthFromOS, _TEXT

--- a/src/coreclr/pal/src/include/pal/context.h
+++ b/src/coreclr/pal/src/include/pal/context.h
@@ -1622,21 +1622,6 @@ DWORD CONTEXTGetExceptionCodeForSignal(const siginfo_t *siginfo,
 
 #endif  // HAVE_MACH_EXCEPTIONS else
 
-#if defined(HOST_ARM64)
-/*++
-Function :
-    CONTEXT_GetSveLengthFromOS
-
-    Gets the SVE vector length
-Parameters :
-    None
-Return value :
-    The SVE vector length in bytes
---*/
-DWORD64
-CONTEXT_GetSveLengthFromOS(
-    );
-#endif // HOST_ARM64
 
 #ifdef __cplusplus
 }

--- a/src/coreclr/pal/src/thread/context.cpp
+++ b/src/coreclr/pal/src/thread/context.cpp
@@ -896,7 +896,11 @@ void CONTEXTToNativeContext(CONST CONTEXT *lpContext, native_context_t *native)
         if (sve && sve->head.size >= SVE_SIG_CONTEXT_SIZE(sve_vq_from_vl(sve->vl)))
         {
             //TODO-SVE: This only handles vector lengths of 128bits.
-            if (CONTEXT_GetSveLengthFromOS() == 16)
+            // Use sve->vl from the signal frame instead of CONTEXT_GetSveLengthFromOS()
+            // (which executes the SVE rdvl instruction) to avoid SIGILL on platforms
+            // that provide an SVE context record without supporting SVE instructions
+            // (e.g. Apple M4 with SME streaming SVE under Virtualization.Framework).
+            if (sve->vl == 16)
             {
                 _ASSERT((lpContext->XStateFeaturesMask & XSTATE_MASK_ARM64_SVE) == XSTATE_MASK_ARM64_SVE);
 
@@ -1255,7 +1259,11 @@ void CONTEXTFromNativeContext(const native_context_t *native, LPCONTEXT lpContex
         if (sve && sve->head.size >= SVE_SIG_CONTEXT_SIZE(sve_vq_from_vl(sve->vl)))
         {
             //TODO-SVE: This only handles vector lengths of 128bits.
-            if (CONTEXT_GetSveLengthFromOS() == 16)
+            // Use sve->vl from the signal frame instead of CONTEXT_GetSveLengthFromOS()
+            // (which executes the SVE rdvl instruction) to avoid SIGILL on platforms
+            // that provide an SVE context record without supporting SVE instructions
+            // (e.g. Apple M4 with SME streaming SVE under Virtualization.Framework).
+            if (sve->vl == 16)
             {
                 _ASSERTE((sve->vl > 0) && (sve->vl % 16 == 0));
                 lpContext->Vl  = sve->vl;

--- a/src/coreclr/pal/src/thread/context.cpp
+++ b/src/coreclr/pal/src/thread/context.cpp
@@ -904,7 +904,10 @@ void CONTEXTToNativeContext(CONST CONTEXT *lpContext, native_context_t *native)
             {
                 _ASSERT((lpContext->XStateFeaturesMask & XSTATE_MASK_ARM64_SVE) == XSTATE_MASK_ARM64_SVE);
 
-                uint16_t vq = sve_vq_from_vl(lpContext->Vl);
+                // Derive vq from the signal frame's vl (the authoritative layout)
+                // rather than lpContext->Vl, to ensure offset calculations always
+                // match the actual frame even in non-debug builds.
+                uint16_t vq = sve_vq_from_vl(sve->vl);
 
                 // Vector length should not have changed.
                 _ASSERTE(lpContext->Vl == sve->vl);
@@ -1265,7 +1268,6 @@ void CONTEXTFromNativeContext(const native_context_t *native, LPCONTEXT lpContex
             // (e.g. Apple M4 with SME streaming SVE under Virtualization.Framework).
             if (sve->vl == 16)
             {
-                _ASSERTE((sve->vl > 0) && (sve->vl % 16 == 0));
                 lpContext->Vl  = sve->vl;
 
                 uint16_t vq = sve_vq_from_vl(sve->vl);

--- a/src/coreclr/pal/src/thread/context.cpp
+++ b/src/coreclr/pal/src/thread/context.cpp
@@ -896,9 +896,8 @@ void CONTEXTToNativeContext(CONST CONTEXT *lpContext, native_context_t *native)
         if (sve && sve->head.size >= SVE_SIG_CONTEXT_SIZE(sve_vq_from_vl(sve->vl)))
         {
             //TODO-SVE: This only handles vector lengths of 128bits.
-            // Use sve->vl from the signal frame instead of CONTEXT_GetSveLengthFromOS()
-            // (which executes the SVE rdvl instruction) to avoid SIGILL on platforms
-            // that provide an SVE context record without supporting SVE instructions
+            // Use sve->vl from the signal frame to avoid SIGILL on platforms that
+            // provide an SVE context record without supporting SVE instructions
             // (e.g. Apple M4 with SME streaming SVE under Virtualization.Framework).
             if (sve->vl == 16)
             {
@@ -1262,12 +1261,12 @@ void CONTEXTFromNativeContext(const native_context_t *native, LPCONTEXT lpContex
         if (sve && sve->head.size >= SVE_SIG_CONTEXT_SIZE(sve_vq_from_vl(sve->vl)))
         {
             //TODO-SVE: This only handles vector lengths of 128bits.
-            // Use sve->vl from the signal frame instead of CONTEXT_GetSveLengthFromOS()
-            // (which executes the SVE rdvl instruction) to avoid SIGILL on platforms
-            // that provide an SVE context record without supporting SVE instructions
+            // Use sve->vl from the signal frame to avoid SIGILL on platforms that
+            // provide an SVE context record without supporting SVE instructions
             // (e.g. Apple M4 with SME streaming SVE under Virtualization.Framework).
             if (sve->vl == 16)
             {
+                _ASSERTE(sve->head.size >= SVE_SIG_CONTEXT_SIZE(sve_vq_from_vl(16)));
                 lpContext->Vl  = sve->vl;
 
                 uint16_t vq = sve_vq_from_vl(sve->vl);

